libsolutil/LazyInit.h

```diff
@@ -0,0 +1,115 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include <libsolutil/Assertions.h>
+#include <libsolutil/Exceptions.h>
+
+#include <memory>
+#include <optional>
+#include <type_traits>
+#include <utility>
+
+namespace solidity::util
+{
+
+DEV_SIMPLE_EXCEPTION(BadLazyInitAccess);
+
+/**
+ * Represents a lazy-initialized value.
+ * @tparam T the type of the lazy-initialized object; may not be a function, reference, array, or void type; may be const-qualified.
+ */
+template<typename T>
+class LazyInit
```

Comment:
I'm a bit bothered by the name and documentation :-).
From those I'd expect this class to actually do lazy initialization, as in:
```
LazyInit<int> x;
x.init([]{ std::cout << "INIT" << std::endl; return 42; }); // It's lazy, so this *doesn't* initialize yet.
*x; // only now the init function is called.
```
So I was first surprised by all the reference captures in the init lambdas, because for such actual lazy initialization those would be pretty problematic...

Do you agree with that and does anyone have a better suggestion for a name?

Reply:
The problem is that we do not want to mention the function that performs the init as part of the declaration. I would be fine with forcing mentioning the function with every access, so basicalyl remove the `*` operator and make `init` return the value. The problem with that approach is that we always need another wrapper function to retrieve the value...

Reply:
I'm not necessarily saying we should go for "real" lazy init - I'm rather saying that if we don't we should look for a more suitable name for it :-).

Reply:
`DelayedInit`? `InitIfNeeded`?

Reply:
``DelayedInit`` would also suggest that ``init`` doesn't really initialize right away... ``InitIfNeeded`` maybe... or maybe just ``Cached``? And call the member ``init`` rather ``initIfNeeded`` or ``initOnce``? I'm not sure. I'd also be fine with just staying with ``LazyInit``, but make the behaviour clearer in the comment, if we don't find any alternative that's really good.

Reply:
Yeah, maybe let's just stick with LazyInit :)